File: app/src/main/java/com/example/practicalibrosreciclerview/Model/DataAdapter.java
package com.example.practicalibrosreciclerview.Model;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import com.example.practicalibrosreciclerview.R;
import com.squareup.picasso.Picasso;
import java.util.ArrayList;
import androidx.annotation.NonNull;
import androidx.cardview.widget.CardView;
import androidx.recyclerview.widget.RecyclerView;

public class DataAdapter
        extends RecyclerView.Adapter<DataAdapter.ViewHolderData>
        implements View.OnClickListener
{
    private ArrayList<Book> bookList;
    private Context context;
    private View.OnClickListener listener;

    public DataAdapter(ArrayList<Book> bookList) {
        this.bookList = bookList;
    }

    @NonNull
    @Override
    public ViewHolderData onCreateViewHolder(@NonNull ViewGroup parent, int viewType)
    {
        context= parent.getContext();
        View itemView = LayoutInflater.from(context)
                .inflate(R.layout.cv_item_book,null,false);

        itemView.setOnClickListener(this);

        return new ViewHolderData(itemView);
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolderData holder, int position)
    {
        holder.dataAsign(bookList.get(position));
    }

    @Override
    public int getItemCount() {
        return bookList.size();
    }

    public void setOnClickListener(View.OnClickListener listener)
     {
         this.listener=listener;
     }

    @Override
    public void onClick(View view)
    {
        if(listener!=null)
        {
            listener.onClick(view);
        }
    }

    public class ViewHolderData extends RecyclerView.ViewHolder
    {
        ImageView imageBook ;
        TextView bookTitle;
        TextView authorName;
        TextView category;


        public ViewHolderData(@NonNull View itemView)
        {
            super(itemView);

            imageBook = itemView.findViewById(R.id.iv_book);
            bookTitle = itemView.findViewById(R.id.tV_book_title);
            authorName= itemView.findViewById(R.id.tV_author_name);
            category= itemView.findViewById(R.id.tV_category);


        }

        public void dataAsign(Book book)
        {
            Picasso.with(context).load(book.getUrlImage()).into(imageBook);
            bookTitle.setText(book.getTitle());
            authorName.setText(book.getAuthor());
            category.setText(book.getCategory());
        }
    }
}

File: app/src/main/java/com/example/practicalibrosreciclerview/model/DataAdapter.kt
package com.example.practicalibrosreciclerview.model

import android.view.LayoutInflater
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import android.widget.Filter
import android.widget.Filterable
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.example.practicalibrosreciclerview.MyApp
import com.example.practicalibrosreciclerview.R
import com.example.practicalibrosreciclerview.model.DataAdapter.ViewHolderData
import com.squareup.picasso.Picasso
import kotlin.collections.ArrayList

class DataAdapter(val bookList: ArrayList<Book>) : RecyclerView.Adapter<ViewHolderData>(), View.OnClickListener, Filterable, OnLongClickListener {

    private lateinit var bookListFull: ArrayList<Book>
    private var listener: View.OnClickListener? = null
    private var longListener: OnLongClickListener? = null

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolderData {
        bookListFull = ArrayList<Book>()
        val context = parent.context
        val itemView = LayoutInflater.from(context).inflate(R.layout.cv_item_book, parent, false)
        itemView.setOnClickListener(this)
        itemView.setOnLongClickListener(this)
        return ViewHolderData(itemView)
    }

    override fun onBindViewHolder(holder: ViewHolderData, position: Int) {
        holder.dataAsign(bookList[position])
    }

    override fun getItemCount(): Int {
        return bookList.size
    }

    override fun onClick(view: View) {
        if (listener != null) {
            listener!!.onClick(view)
        }
    }

    override fun onLongClick(view: View): Boolean {
        if (longListener != null) {
            longListener!!.onLongClick(view)
        }
        return true
    }

    override fun getFilter(): Filter {
        return exampleFilter
    }
    fun setOnClickListener(listener: View.OnClickListener?) {
        this.listener = listener
    }

    fun setOnLongClickListener(listener: OnLongClickListener?) {
        longListener = listener
    }



    private val exampleFilter: Filter = object : Filter() {
        var results = FilterResults()

        override fun performFiltering(constraint: CharSequence): FilterResults {
            val filteredList: MutableList<Book> = ArrayList()
            if (constraint.isEmpty()) {
                filteredList.addAll(bookListFull)
            } else {
                val filterPattern = constraint.toString().trim { it <= ' ' }
                for (item in bookListFull) {
                    if (item.title.contains(filterPattern)
                            || item.isbn.contains(filterPattern)) {
                        filteredList.add(item)
                    }
                }
            }
            results.values = filteredList
            return results
        }

        override fun publishResults(charSequence: CharSequence, filterResults: FilterResults) {
            bookList.clear()

            if (filterResults.values is ArrayList<*>){
                bookList.addAll(filterResults.values as ArrayList<Book>)

            }
            notifyDataSetChanged()
        }
    }

    inner class ViewHolderData(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private var imageBook: ImageView = itemView.findViewById(R.id.iv_book)
        private var bookTitle: TextView = itemView.findViewById(R.id.tV_book_title)
        private var authorName: TextView = itemView.findViewById(R.id.tV_author_name)
        private var category: TextView = itemView.findViewById(R.id.tV_category)

        fun dataAsign(book: Book) {
            Picasso.get().load(book.imageUrl).into(imageBook)
            bookTitle.text = MyApp.context!!.getString(R.string.book_title, book.title)
            authorName.text = MyApp.context!!.getString(R.string.author_name, book.authorFirstName, book.authorLastName)
            category.text = MyApp.context!!.getString(R.string.category, book.category)
        }

    }

    init {
        bookListFull = ArrayList(bookList)
    }
}